### ISSUE TITLE
Generic type not propagated/infered correctly to callback 

### ISSUE THREAD

**Issue by Andarist** [microsoft/TypeScript#31933]
**TypeScript Version:**  3.5.2

**Code**

```ts
type ActionFunction<TContext> = (context: TContext) => void

interface ActionObject<TContext> {
    type: string;
    exec?: ActionFunction<TContext>;
}

interface MachineConfig<TContext> {
    context?: TContext;
    entry?: ActionObject<TContext>;
}

declare function Machine<TContext>(
    config: MachineConfig<TContext>
): void;

type Assigner<TContext> = (context: TContext) => Partial<TContext>;

interface AssignAction<TContext> extends ActionObject<TContext> {
    type: "xstate.assign";
    assignment: Assigner<TContext>;
}

declare function assign<TContext>(
    assignment: Assigner<TContext>
): AssignAction<TContext>

Machine<{ counter: number; serverRef?: boolean }>({
    context: { counter: 0, serverRef: undefined },
    entry: assign(() => ({
        serverRef: true
    }))
});
```

**Expected behavior:**

This should type check correctly. The problem here seems that generic is not propagated to that callback correctly - if you add an argument to that callback (which stays unused!) then it type checks just fine.

It's somehow connected to the fact that `assign` returns a subtype of `ActionObject` which has generic optional `exec` property.

**Actual behavior:**

It doesn't type check

**Playground Link:** 
[link](https://www.typescriptlang.org/play/index.html#src=type%20ActionFunction%3CTContext%3E%20%3D%20(context%3A%20TContext)%20%3D%3E%20void%0D%0A%0D%0Ainterface%20ActionObject%3CTContext%3E%20%7B%0D%0A%20%20%20%20type%3A%20string%3B%0D%0A%20%20%20%20exec%3F%3A%20ActionFunction%3CTContext%3E%3B%0D%0A%7D%0D%0A%0D%0Ainterface%20MachineConfig%3CTContext%3E%20%7B%0D%0A%20%20%20%20context%3F%3A%20TContext%3B%0D%0A%20%20%20%20entry%3F%3A%20ActionObject%3CTContext%3E%3B%0D%0A%7D%0D%0A%0D%0Adeclare%20function%20Machine%3CTContext%3E(%0D%0A%20%20%20%20config%3A%20MachineConfig%3CTContext%3E%0D%0A)%3A%20void%3B%0D%0A%0D%0Atype%20Assigner%3CTContext%3E%20%3D%20(context%3A%20TContext)%20%3D%3E%20Partial%3CTContext%3E%3B%0D%0A%0D%0Ainterface%20AssignAction%3CTContext%3E%20extends%20ActionObject%3CTContext%3E%20%7B%0D%0A%20%20%20%20type%3A%20%22xstate.assign%22%3B%0D%0A%20%20%20%20assignment%3A%20Assigner%3CTContext%3E%3B%0D%0A%7D%0D%0A%0D%0Adeclare%20function%20assign%3CTContext%3E(%0D%0A%20%20%20%20assignment%3A%20Assigner%3CTContext%3E%0D%0A)%3A%20AssignAction%3CTContext%3E%0D%0A%0D%0AMachine%3C%7B%20counter%3A%20number%3B%20serverRef%3F%3A%20boolean%20%7D%3E(%7B%0D%0A%20%20%20%20context%3A%20%7B%20counter%3A%200%2C%20serverRef%3A%20undefined%20%7D%2C%0D%0A%20%20%20%20entry%3A%20assign(()%20%3D%3E%20(%7B%0D%0A%20%20%20%20%20%20%20%20serverRef%3A%20true%0D%0A%20%20%20%20%7D))%0D%0A%7D)%3B%0D%0A%0D%0A)

**Comment by dragomirtitian** [comment]
I think typescript will favor inferring the type parameter from arguments rather then infer it based on return type:

```ts
function id<T>(o : T) : T {
    return o;
}
let o : { foo: number, bar: string} = id({ foo: 1 }); // T is { foo: number }

function produce<T>() : T {
    return null!;
}
let p: { foo: number, bar: string} = produce(); // T is inferred to  { foo: number, bar: string} (you can see this on hover)
```

A workaround might be to make it impossible to infer form argument, in effect forcing the compiler to infer based on return type:

```ts
type ActionFunction<TContext> = (context: TContext) => void

interface ActionObject<TContext> {
    type: string;
    exec?: ActionFunction<TContext>;
}

interface MachineConfig<TContext> {
    context?: TContext;
    entry?: ActionObject<TContext>;
}

declare function Machine<TContext>(
    config: MachineConfig<TContext>
): void;

type Assigner<TContext> = (context: TContext) => Partial<TContext>;

interface AssignAction<TContext> extends ActionObject<TContext> {
    type: "xstate.assign";
    assignment: Assigner<TContext>;
}

declare function assign<T extends ActionObject<any>>(
    assignment: T extends ActionObject<infer TContext> ? Assigner<TContext>: never
): T

Machine<{ counter: number; serverRef?: boolean }>({
    context: { counter: 0, serverRef: undefined },
    entry: assign(() => ({
        serverRef: true
    }))
});

Machine<{ counter: number; serverRef?: boolean }>({
    context: { counter: 0, serverRef: undefined },
    entry: assign((c) => ({ // c: { counter: number; serverRef?: boolean; }
        serverRef: true
    }))
});

```

**Comment by Andarist** [comment]
@dragomirtitian Thanks for the reply! I definitely need to digest this more thoroughly as it might give me a better understanding of TS anyway - but your corrected version doesn't yield my expected result. `c` argument gets inferred to `any`.

**Comment by dragomirtitian** [comment]
Hm.. strange, the version above works (infers the correct `T` ) if `strictNullChecks` are disabled ..

**Comment by dragomirtitian** [comment]
The code above worked if `strictNullChecks` are disabled, to make it work under `strictNullChecks`  the declaration for `assign`  must be 
```ts
declare function assign<T extends ActionObject<any> | undefined>
```

**Comment by Andarist** [comment]
If you are interested I've ended up with this - https://github.com/davidkpiano/xstate/pull/503 .

I didn't want to have T extending ActionObject because `assign` should be able to accept explicit type argument for `TContext`, nevertheless I've managed to force TS to infer this correctly by a simple conditional type for `assignment`: 
```ts
TContext extends any ?
    (Assigner<TContext, TEvent> | PropertyAssigner<TContext, TEvent>)
    : never
```

I'm still unsure why inferring from return type makes the problem go away as the return value is not providing all matching properties of `TContext` (it doesnt explicitly disallow them though). Does it mean that TS in fact check there for what could be inferred and checks if it can "widen" that return type to the "expected" one? How this all resolves step by step is kinda blurry to me.
